adapter/slack.py

```diff
@@ -1,79 +1,112 @@
+        engine_selector: ChatEngineSelector,
+        bot_service: BotService,
+    ) -> None:
         self.app = app
-        self.chatbot = chatbot
-        self.bot_controller = bot_controller
+        self.engine_selector = engine_selector
+        self.bot_service = bot_service
         self.handler = SlackRequestHandler(self.app)
         self.logger = logger.bind(service="SlackAdapter")
 
     # Function to listen for events on Slack. No need to test this since this is purely dependent on
     # Bolt
-    # pragma: no cover
-    async def handle_events(self, req: Request):
+    async def handle_events(self, req: Request):  # pragma: no cover
         return await self.handler.handle(req)
 
     async def ask(self, request: Request):
         data = await request.form()
-    
+
         channel_id = data.get("channel_id")
         user_id = data.get("user_id")
         parameter = data.get("text", "")
-          
-        parts = parameter.split(" ", 1) 
+
+        parts = parameter.split(" ", 1)
         if len(parts) == 2:
-            _, question = parts
+            bot_id, question = parts
         else:
-            raise HTTPException(status_code=400, detail="Missing parameter in the request.")
-        
-        question = f"<@{user_id}> asked: \n\n\"{question}\" "
+            return {
+                "text": "Missing parameter in the request.",
+            }
+
+        question = f'<@{user_id}> asked: \n\n"{question}" '
 
         try:
-            response = self.app.client.chat_postMessage(channel=channel_id, text=question)
+            chatbot = self.bot_service.get_chatbot_by_id(bot_id=bot_id)
+
+            if chatbot is None:
+                return {"text": "Whoops. Can't found the chatbot you're looking for."}
+
+            response = self.app.client.chat_postMessage(
+                channel=channel_id, text=question
+            )
             thread_ts = response["ts"]
-            
+
             loading_message = self.app.client.chat_postMessage(
                 channel=channel_id,
                 text=":hourglass_flowing_sand: Processing your request, please wait...",
-                thread_ts=thread_ts
+                thread_ts=thread_ts,
             )
-            
-            chatbot_response = self.chatbot.generate_response(query=question)
-            
+
+            bot_engine = self.engine_selector.select_engine(engine_type=chatbot.model)
+
+            chatbot_response = bot_engine.generate_response(query=question)
+
             self.app.client.chat_update(
-                channel=channel_id,
-                ts=loading_message["ts"],
-                text=chatbot_response
+                channel=channel_id, ts=loading_message["ts"], text=chatbot_response
             )
-            
+
             return Response(status_code=200)
 
+        except sqlalchemy.exc.DataError as e:  # pragma: no cover
+            return {"text": "Whoops. Can't found the chatbot you're looking for"}

```

Comment:
I think this error message needs to be differentiated with the one above in line 52. The one in line 52 happens if there really is no bot with the id, while this line happens because of some error. Maybe something like "Database Error" or something.

Reply:
That one is to handle non-UUIDs as the database throws an error when that happens (e.g. 1212121). Since the error we received is quite hard to parse, I used this as a catch-all for now

Reply:
understandable, thanks for the explanation